src/lowlevel/pin_util.rs

```diff
@@ -0,0 +1,91 @@
+use std::ops::Deref;
+use std::pin::Pin;
+use std::sync::Arc;
+
+use tokio::sync::{Mutex as MutexAsync, MutexGuard as MutexAsyncGuard};
+
+#[derive(Debug)]
+pub(super) struct PinnedArc<T>(Arc<T>);
+
+impl<T> Clone for PinnedArc<T> {
+    fn clone(&self) -> Self {
+        Self(self.0.clone())
+    }
+}
+
+impl<T> PinnedArc<T> {
+    pub(super) fn new(val: T) -> Self {
+        Self(Arc::new(val))
+    }
+
+    pub(super) fn strong_count(arc: &Self) -> usize {
+        Arc::strong_count(&arc.0)
+    }
+
+    pub(super) fn get_mut(arc: &mut Self) -> Option<&mut T> {
+        Arc::get_mut(&mut arc.0)
+    }
+
+    pub(super) fn get_pinned_mut(arc: &mut Self) -> Option<Pin<&mut T>> {
+        // Safety:
+        //
+        // Arc is essentially ref-counted box, which provides stable address for T,
+        // moving the arc does not change the address of T.
+        Arc::get_mut(&mut arc.0).map(|val| unsafe { Pin::new_unchecked(val) })
```

Comment:
I'm pretty sure this isn't sound. For example:
```rust
let mut x = PinnedArc::new(Vec::new());
takes_pin(PinnedArc::get_pinned_mut(&mut x).unwrap());
// at this point it is absolutely positively not legal to move the `Vec`, since we've given out `Pin<&mut >>` to it.
// and yet:
let was_x = std::mem::replace(PinnedArc::get_mut(&mut x).unwrap(), Vec::new());
```

Reply:
Looks like the problem is `PinnedArc::get_mut`.

The problem should be fixed if I removed this function.

Reply:
I have removed `PinnedArc::get_mut`, now it should be sound.